tests/functional/behave_features/common/utils/owner_file_submission.py

```diff
@@ -0,0 +1,351 @@
+# -*- coding: utf-8 -*-
+"""Utility class for setting up and manipulating owner file submissions"""
+
+import os
+import logging
+import pathlib
+import uuid
+import git
+import json
+
+from dataclasses import dataclass
+from tempfile import TemporaryDirectory
+from string import Template
+from pathlib import Path
+
+from common.utils.workflow_repo_manager import WorkflowRepoManager
+from common.utils.secret import E2ETestSecretOneShot
+from common.utils.set_directory import SetDirectory
+from common.utils.e2e_templates import base_owners_file
+import common.utils.github as github
+import common.utils.env as env
+import common.utils.setttings as settings
+
+
+@dataclass
+class OwnersFileSubmissionsE2ETest:
+    # Generated at initialization. Identifies each instance uniquely.
+    # Used for PRs titles, branch names, etc.
+    uuid: str = ""
+    head_sha: str = ""
+
+    test_name: str = ""  # Meaningful test name for this test, displayed in PR title
+    # test_report: str = ''
+    # chart_directory: str = ''
+    secrets: E2ETestSecretOneShot = E2ETestSecretOneShot()
+    old_cwd: str = os.getcwd()
+    # This is the worktree for this test run. To be used for context management where applicable.
+    temp_dir: TemporaryDirectory = None
+    # Whether this is being executed in GitHub Actions.
+    github_actions: str = os.environ.get("GITHUB_ACTIONS")
+    # The name of the chart as supplied by the user.
+    chart_base_name = ""
+    # The name of the chart as supplied by the user appended with a unique identifier.
+    chart_name = ""
+    # The vendor name, e.g. 'redhat'
+    vendor_label = ""
+    # The vendor category, e.g. 'partner'
+    vendor_category = ""
+    # The vendor's pretty name, e.g. 'Red Hat'
+    vendor_name = ""
+    # The PR number in string format for this test run.
+    pull_request: str = ""
+
+    # Manages the local repository, branches, worktress, etc.
+    # and facilitates pushes to remotes.
+    repo_manager: WorkflowRepoManager = None
+    repo: git.Repo = None
+
+    # Combines this instance's unique ID + the repository base branch hash
+    branch_base_id: str = None
+
+    def __post_init__(self) -> None:
+        """Generates identifiers for an instance to use."""
+        logging.debug("RedHatOwnersFileSubmissionE2ETest --> __post_init__ called!")
+        # New data for this instance.
+        self.uuid = uuid.uuid4().hex
+        self.repo_manager = WorkflowRepoManager()
```

Comment:
Setting default values at the class level produces the same class value being set if the value isn't set when the instance is created.

E.g.

```python
import uuid

class CustomClass:
    x: str = uuid.uuid4().hex
    y: str = None

    def __init__(self):
        self.y = uuid.uuid4().hex


i = CustomClass()
print(f'i.x is {i.x} | i.y is {i.y}')
a = CustomClass()
print(f'a.x is {a.x} | a.y is {a.y}')
```

Produces:

```
i.x is dad783e028914353ae9f703d3bfbc9e6 | i.y is 9213b42f0ea84bff9a00e7ba43e01195
a.x is dad783e028914353ae9f703d3bfbc9e6 | a.y is 39641c7f6021416186f0bb0608d9d325
```

`i` and `a` have `x` properties with the same value because new ones weren't passed in. The initialization happens only once per class. By setting the value to what we want in the initialization, we force unique values for each instance.

In theory, I would want this value to be immutable, too, but I didn't feel like putting a bunch of dunder variables in place.

Reply:
You're right. In that case one can use `default_factory`:

```py
import uuid

from dataclasses import dataclass, field

@dataclass
class CustomClass:
    x: str = uuid.uuid4().hex
    y: str = None
    z: str = field(default_factory=lambda: uuid.uuid4().hex)

    def __post_init__(self): # moved to __post_init__
        self.y = uuid.uuid4().hex


i = CustomClass()
print(f'i.x is {i.x} | i.y is {i.y} | i.z is {i.z}')
a = CustomClass()
print(f'a.x is {a.x} | a.y is {a.y} | a.z is {a.z}')
```

which gives:

```
i.x is 9edba68c12fa45c6bbdfe4f54b796aca | i.y is 3983618d6858415ab475e38207f82bed | i.z is 6d050633f11341088ae7663816bc3c68
a.x is 9edba68c12fa45c6bbdfe4f54b796aca | a.y is bb6832743b5447cdb39fe19454e5c131 | a.z is ec1effef4d44488d8e23f6273fe30a19
```

Reply:
Had a few minutes to read up on Dataclasses and FIelds, today. I don't quite thing dataclasses are fit for purpose. I believe their goal is to simplify classes that don't contain a large amount of methods/logic, but rather contain a series of fields. In other words, they reduce the boilerplate for writing classes that end up recieving (as params) or generating (e.g. using default_factory) their values.

All that said, I used a dataclass here to maintain consistency. To that end, default_factory is just as effective a solution as what was present. And I tested `init=False` as well, which just prevents the field from being set. The combination satisfies the need - unique values per instance.